lighthouse-core/computed/metrics/lantern-first-contentful-paint.js

```diff
@@ -24,52 +26,152 @@ class LanternFirstContentfulPaint extends LanternMetric {
+    // so we always include those CPU nodes.
+    const firstLayout = cpuNodes.find(node => node.didPerformLayout());
+    if (firstLayout) blockingCpuNodeIds.add(firstLayout.id);
+    const firstPaint = cpuNodes.find(node => node.childEvents.some(e => e.name === 'Paint'));
+    if (firstPaint) blockingCpuNodeIds.add(firstPaint.id);
+    const firstParse = cpuNodes.find(node => node.childEvents.some(e => e.name === 'ParseHTML'));
+    if (firstParse) blockingCpuNodeIds.add(firstParse.id);
+
+    // If a CPU filter was passed in, we also want to include those extra nodes.
+    if (extraBlockingCpuNodesToIncludeFilter) {
+      cpuNodes
+        .filter(extraBlockingCpuNodesToIncludeFilter)
+        .forEach(node => blockingCpuNodeIds.add(node.id));
+    }
+
+    return {
+      definitelyNotRenderBlockingScriptUrls,
+      blockingCpuNodeIds,
+    };
   }
 
   /**
+   * This function computes the graph required for the first paint of interest.
+   *
    * @param {Node} dependencyGraph
-   * @param {LH.Artifacts.TraceOfTab} traceOfTab
+   * @param {number} paintTs
+   * @param {function(NetworkNode):boolean} blockingScriptFilter
+   * @param {function(CPUNode):boolean=} extraBlockingCpuNodesToIncludeFilter
    * @return {Node}
    */
-  static getPessimisticGraph(dependencyGraph, traceOfTab) {
-    const fcp = traceOfTab.timestamps.firstContentfulPaint;
-    const blockingScriptUrls = LanternMetric.getScriptUrls(dependencyGraph, node => {
-      return node.endTime <= fcp && node.hasRenderBlockingPriority();
-    });
+  static getFirstPaintBasedGraph(
+      dependencyGraph,
+      paintTs,
+      blockingScriptFilter,
+      extraBlockingCpuNodesToIncludeFilter
+  ) {
+    const {
+      definitelyNotRenderBlockingScriptUrls,
+      blockingCpuNodeIds,
+    } = this.getBlockingCpuData(
+      dependencyGraph,
+      paintTs,
+      blockingScriptFilter,
+      extraBlockingCpuNodesToIncludeFilter
+    );
 
     return dependencyGraph.cloneWithRelationships(node => {
-      if (node.endTime > fcp && !node.isMainDocument()) return false;
-      // Include EvaluateScript tasks for blocking scripts
-      if (node.type === BaseNode.TYPES.CPU) {
-        return node.isEvaluateScriptFor(blockingScriptUrls);
-      }
+      if (node.type === BaseNode.TYPES.NETWORK) {
+        // Exclude all nodes that ended after FCP (except for the main document which we always consider necessary)
+        if (node.endTime > paintTs && !node.isMainDocument()) return false;
 
-      // Include non-script-initiated network requests with a render-blocking priority
-      return node.hasRenderBlockingPriority();
+        const url = node.record.url;
+        // If the URL definitely wasn't render-blocking then we filter it out.
+        if (definitelyNotRenderBlockingScriptUrls.has(url)) {
+          return false;
+        }
+        return node.hasRenderBlockingPriority();
```

Comment:
this line seems to get into a double negative situation, though maybe I'm just reading wrong here.

The optimistic graph includes a filter for `node => node.hasRenderBlockingPriority() && node.initiatorType !== 'script'`, which means a node with render blocking priority but isn't `initiatorType` `script` is not considered possibly blocking.

That means it can't end up in `definitelyNotRenderBlockingScriptUrls`, which means the return value falls to this line, and since the node has render blocking priority it'll return true and be included as blocking, even though the filter was saying it shouldn't be considered blocking.

Is that not intended (and this line should maybe be `return blockingScriptFilter(node)`?) or it is intended and we should do some renaming or something to make it clearer?

Reply:
yes great catch!! and now we improve the metrics even more, huzzah 🎉 :D